### PR TITLE
Standardise reloading ammo [WIP] [CR]

### LINKED ISSUE #16350: Mismatch between stack size of casing/primer and stack size of product in reloaded ammo crafting
A lot of them dont add up.
For example: 
reloaded .30-06 Springfield require 10 casings to craft 20 bullets.
reloaded .44 FMJ require 50 casings to craft 20 bullets.

It would be nice to extract the casing/primer requirement from the stack size of end product dynamically.

PS. some item names also dont line up. We have "handloaded" .22 FMJ and "reloaded" .44 FMJ.

### PR DESCRIPTION
**Pull request by mugling**
### Component requirements
- New recipes are per-cartridge
- Will fix #16350 and #16849
### Skill requirements
- Base requirements depend upon weapon skill (eg. `reloaded_9mm` requires `pistol 1`)
- Less common variants are higher level (eg. `reloaded_9mmP2` requires `pistol 4`)
- Intended range is from levels 1-6 with most ammo with only a few rare types above level 4
### Book requirements
- Weapon manuals provides relevant recipes at the skill level needed to craft them
- Weapon magazines provide no recipes
- Handloading manual provides recipes for all skills but at one level about corresponding manual
- Top tier ammo (`reloaded_9mmP2`, `38_super,`45_super`) is excluded from handloading manual
- Handloading manual no longer requires nor trains fabrication
- SMG manuals provide appropriate pistol ammo

Nett effect of the above is that `9mm` ammo is very easy to craft with further types gradually increasing in difficulty from there. Combined with the recent changes preventing reading to learn recipes the handloading manual now functions correctly as a useful but bulky reference book.
### Time requirements
- 5 minutes per round but with very significant bonuses when crafting in batches
- Disassembly is 30 seconds/round with no bonus
### Comments requested
- Merge strategy (single PR, per ammo-type or per n-LoC)
- Above balance outline

### CODE REVIEW

**Comment by Vidiveni** [comment]
I like replacing fabrication as a requirement, but I'm concerned with having to train pistols to be able to craft ammo for my SMG.

**Comment by mugling** [comment]
> but I'm concerned with having to train pistols to be able to craft ammo for my SMG.

See preamble

> SMG manuals provide appropriate pistol ammo

**Comment by Vidiveni** [comment]
Reloaded 9mm recipe says "skill_used": "pistol". Sorry if I misunderstand.

**Comment by mugling** [comment]
No actually you are correct. That's an intrinsic limitation of the current crafting system

**Comment by Reclusive-reptile** [comment]
The use and maintenance of a firearm does not necessarily imply the knowledge and skills required to produce said firearm's ammunition. It is probably best to keep reloading ammo in the fabrication skill. As for the manuals containing the  higher tier loads rather than the hand loader's helper, the reverse is likely the more thematically consistent approach.

**Comment by mugling** [comment]
> The use and maintenance of a firearm does not necessarily imply the knowledge and skills required to produce said firearm's ammunition.

No, but there is probably an association and it is the closest approximation we have.

> It is probably best to keep reloading ammo in the fabrication skill.

Too many recipes use `fabrication`. Being able to reload ammunition has even less to do with being able to craft vehicle parts, armor and tooling. Presently in our skill system `fabrication` is one of the worst offenders for both grinding and poor spacing of recipes.

> As for the manuals containing the higher tier loads

As @Coolthulhu pointed out elsewhere books are one of the worst aspects of the game. @kevingranade's recent PR removing memorisation of recipes via reading helped but the current progression still is grind lockpicks, read two books, craft anything on day 3.

The handloaders helper shouldn't grant fabrication and the skill manuals should actually have some reward considering it's both trivially easy to gain level 3 weapon skill yet those skills are not especially useful overall.

The suggestion elsewhere of an achievements system is very appealing as it rewards gameplay and not grinding. This could actually work well for ammo, eg. make good shot at long range and unlock `556` but for now this PR should suffice.

**Comment by Coolthulhu** [comment]
Ranged weapon skills in general cover aiming. So far it was always marksmanship or mechanics which covered gun maintenance.

Fabrication does make sense here because it is the generic crafting skill for things that do not move.

Limiting recipe learning to high gun skills sounds good, but having reloading train gun skills (and waste resources when those are too low) sounds rather weird to me. It would make more sense if it was the gun skill limiting recipe learning, but some other skill governing the crafting.

I'm not sure which option is best here.

File: src/crafting.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ void load_recipe( JsonObject &jsobj, const std::string & /* src */, bool uncraft
     std::string result = jsobj.get_string( "result" );
     std::string category = uncraft ? "CC_NONCRAFT" : jsobj.get_string( "category" );
     int time = jsobj.get_int( "time" );
-    int difficulty = jsobj.get_int( "difficulty" );
+    int difficulty = uncraft ? 0 : jsobj.get_int( "difficulty" );
```
Comment (Coolthulhu) [inline src/crafting.cpp]: Uncraft recipes can be difficult. This makes sense when the object is hard to dismantle properly - for example a robot.

**Comment by Vidiveni** [comment]
Agree on the points, but it would feel a bit bad to play with SMG's after this PR unless we make duplicate recipes for all dual or triple ammo types (if any). 

I would honestly be fine with having the ammo list bloated by duplicates this way.

**Comment by mugling** [comment]
> Ranged weapon skills in general cover aiming. So far it was always marksmanship or mechanics which covered gun maintenance.

Marksmanship should be aiming. Weapon skill should provide a small bonus but mostly affect reloading, maintenance and what ammo/mods can be crafted. This results in advancement through doing as opposed to via grinding crafts.

> Limiting recipe learning to high gun skills sounds good, but having reloading train gun skills (and waste resources when those are too low) sounds rather weird to me. It would make more sense if it was the gun skill limiting recipe learning, but some other skill governing the crafting.

That could actually work if we split requirements into those that are needed and those that affect success rate?

**Comment by cainiaowu** [comment]
Arrow need the same single craft standardisation. 

Right now most arrow recipe is based 10.

**Comment by mugling** [comment]
@cainiaowu noted

**Comment by Griffinhart** [comment]
Seems to me like linking bullet crafting to weapon skills would be analogous to linking vehicle part crafting to driving.

Knowing how to shoot a gun/drive a car is somewhat different from knowing how to handload a bullet/make a gas tank.

**Comment by mugling** [comment]
> Knowing how to shoot a gun/drive a car is somewhat different from knowing how to handload a bullet/make a gas tank.

Why? For example how would you determine what the best place to fit an axle without practice operating a range of vehicles in differing circumstances?

**Comment by Griffinhart** [comment]
You are conflating "making an axle" with "building a car".

The intrinsic ballistic properties of a 9mm round are invariant with respect to whether it's being fired from a Glock 19 handgun or an MP5 submachinegun.

Similarly, the intrinsic physical properties of an axle are invariant with respect to whether it's being mounted to a 4-door sedan or a semi-truck.

You could make argument that an axle for use on a 4-door sedan would be different from an axle for use on a semi-truck, and you'd be correct, but a) that's not what's being discussed - "making two different axles for use on two different vehicles" is different from "making one axle for use on two different vehicles" and b) CDDA doesn't model specific different axles regardless; similarly, CDDA doesn't model 9mm-designed-for-handgun differently from 9mm-designed-for-SMG.

**Comment by Soadreqm** [comment]
>Marksmanship should be aiming. Weapon skill should provide a small bonus but mostly affect reloading, maintenance and what ammo/mods can be crafted.

I can sort of see the logic in this, but I don't think it's consistent with the current system. Right now, the individual gun skills refer to the general shape of the gun and how it's wielded, not to its mechanical properties.

If you want to go with something like this, I'd suggest reorganizing the gun skills according to ammo caliber instead of gun size. I've never taken a gun apart, but I've generally assumed that the ammunition used is fairly closely connected to a gun's inner workings. Like, a pistol-caliber rifle has more in common with a pistol than a .50 BGM anti-materiel rifle, and a homemade pistol that loads a single 12-gauge shotgun shell is closer to a shotgun in operation than a .38 revolver. Something like small-caliber, medium-caliber, high-caliber and shot instead of pistols, submachine guns, rifles and shotguns. The ammo skill would cover maintenance, reloading ammo and crafting your own guns, and marksmanship would be general aiming ability.

If that doesn't sound workable, I think it'd be better to just use marksmanship as the generic gun maintenance skill and have the other skills govern aiming. It's already used for installing mods. Having all ammo crafting recipes tied to a single skill that isn't fabrication seems like it would still be an improvement, and it'd solve the problem with having to learn to use guns you don't use so you can craft the ammo you need.

**Comment by mugling** [comment]
> I think it'd be better to just use marksmanship as the generic gun maintenance skill and have the other skills govern aiming. It's already used for installing mods

Apart from the various sights mods actually depend on relevant weapon skill

> I'd suggest reorganizing the gun skills according to ammo caliber instead of gun size

This would lead to an excessive number of skills. Also a 9mm carbine shouldn't train the same skill as a 9mm pistol.

**Comment by Soadreqm** [comment]
>This would lead to an excessive number of skills.

Quite possible, although it depends on how fine you need the categories to be. You might be able to make do with something like
- small bullets
- big bullets
- shotgun shells
- energy weapons
- launchers
- archery
for no increase in the number of skills. At least if you can pigeonhole all the special cases like nail rifles and black-powder guns somewhere. Of course, you'd also lose a whole weapon category from lumping bullets into only two groups.

>Also a 9mm carbine shouldn't train the same skill as a 9mm pistol.

Why not? Assuming that it's not primarily a shooting skill, but a gun maintenance skill. As said, I don't know much about guns, but it seems that the moving parts should be fairly similar for them to be able to use the same caliber.

**Comment by mugling** [comment]
> As said, I don't know much about guns, but it seems that the moving parts should be fairly similar for them to be able to use the same caliber.

We have 9mm revolvers, 9mm pistols, 9mm smg, 9mm rifles. Those all have **very** different moving parts.

> You might be able to make do with something like

That's a complete rebalancing of all the JSON. This PR will close a milestone issue towards `0.D` and I also just finished a large chunk of the aiming code. Both are excellent reasons not to pursue expansive changes/rewrites.

**Comment by Griffinhart** [comment]
>I've never taken a gun apart, but I've generally assumed that the ammunition used is fairly closely connected to a gun's inner workings. Like, a pistol-caliber rifle has more in common with a pistol than a .50 BGM anti-materiel rifle, and a homemade pistol that loads a single 12-gauge shotgun shell is closer to a shotgun in operation than a .38 revolver. 

This is a very incorrect assumption. A .357 revolver basically works nothing like a .357 semi-automatic handgun.

**Comment by Rivet-the-Zombie** [comment]
I don't see any reason why assembling bullets should require firearm-handling skills whatsoever. A person could could theoretically spend their life making bullets and become world-class in that field without ever even handling a firearm in the first place.

Operating [this](http://i.imgur.com/OAwb1kd.jpg) has no relation to operating [this](http://i.imgur.com/CHDQZ2S.jpg).

**Comment by mugling** [comment]
I hope that's not what you keep on your kitchen table

**Comment by mugling** [comment]
On a more serious note one of the aims of moving recipes to partially depend on weapon skills is to improve the reward curve. Currently the player gains all the ammo recipes at the same stage of the game which results in only the top tier ever being used. There isn't much reason to craft `9mm` when you can craft `9mm +P+` etc - this is a shame as someone has taken the time to implement all the various types.

Lack of progression (reward) is bad for gameplay. I'm not averse to other solutions but putting all the recipes in fab 3-4 spoils our differing ammo types

**Comment by mugling** [comment]
I've completed sufficient of the crafting refactor to resume work on closing this issue in preparation for `0.D`. This will use weapon skills for the recipe requirements as outlined above. The only outstanding issue is how to handle SMG cartridges.

Three options:
1. Separate recipes for both `pistol` and `smg` skill
2. Recipes for `pistol` skill only
3. Further crafting refactor allowing alternative (`OR`) skill requirements

I can't see an advantage of (2) and I think (3) would be an excessive delay for `0.D`. In the absence of other suggestion I'm going to implement (1).

**Comment by Reclusive-reptile** [comment]
> I'm going to implement (1).
This still leads to the problem that such a change is unrealistic and muddies the separation of "kill stuff" skills and "make stuff" skills. Pistol caliber rifle carbines are in the game, hand guns that use shot shells are in the game, and one need not bone up in cutting weapons to craft a combat knife. There is no relation between hand loading ammunition and practice with a firearm. Hell, Margolin was blind, never fired a gun in his life, and he was considered one of the best soviet designers of precision firearms.

**Comment by mugling** [comment]
@Reclusive-reptile realism has nothing to do with good gameplay. Try PR'ing a 100x increase in the volume of the handloading equipment (see that picture posted earlier) and see how popular a change that is. Sure it's realistic, but fun, no?

The problem is that there otherwise no reward curve and we might as well have one type of ammo (the best type only). Ideally I want a reward/achievement system that unlocks recipes but that's more tricky to achieve. Having all the ammo recipes available in the early game means we might as well remove 95% of the ammo types (I don't have a problem with that but I supect others won't like that change).

Consider it's a `0.D` issue and decide if we think further lengthy (and possibly endless) discussions are justified in delaying the release given that I've done all the work leading up to this issue and am now ready to close it.

**Comment by Soadreqm** [comment]
For what it's worth, I fully endorse just going with option 1) and being done with it. It fails to account for some of the edge cases like pistol-caliber rifles and shot pistols, and clutters the crafting menu with duplicate recipes, but those aren't big issues. Look into a more elegant solution whenever someone reworks the skills or the crafting system to support it.

**Comment by mugling** [comment]
> Look into a more elegant solution whenever someone reworks the skills or the crafting system to support it.

There is a lot to be said for an actual solution versus endless discussion. Before I write the implementation of (1) does anyone have anything else they'd like to propose.

**Comment by Vidiveni** [comment]
I would be happy to see the handloading recipe 'per-cartridge' fixes get prioritized, even before the other changes are fully finished. I understand it doubles some work, but still. 

Feels like a shame to have those bugs still in the game. Been looking forward to the continuation of this PR-chain because it eventually opens up new sub-playstyles (elephant gun as main weapon anyone?).

**Comment by mugling** [comment]
I'm going to start on this next

**Comment by kevingranade** [comment]
On Aug 7, 2016 2:53 AM, "mugling" <notifications@github.com> wrote:

The use and maintenance of a firearm does not necessarily imply the
knowledge and skills required to produce said firearm's ammunition.

No, but there is probably an association and it is the closest
approximation we have.

It is probably best to keep reloading ammo in the fabrication skill.

Too many recipes use fabrication. Being able to reload ammunition has even
less to do with being able to craft vehicle parts, armor and tooling.
Presently in our skill system fabrication is one of the worst offenders for
both grinding and poor spacing of recipes.

If fabrication has too many recipes, fragment it, don't shoehorn recipes
into skills where they don't fit.

This could either be a totally distinct skill or a sub-skill like
fabrication->reloading where both are trained by practice, fabrication
handles how good you are at it, and reloading mostly meters ability to
execute various recipes.

Weapon-type skills are intended for *handling* the weapon, including
aiming, recoil management, and reloading.
They can be stretched to include weapon repair and application/removal of
mods since these are activities learned together with marksmanship in a
shooting curriculum (so similarly textbooks and magazines), and are aided
by sheer familiarity with the weapon.

Ammo reloading on the other hand is at a huge disconnect from weapon
handling, they aren't taught together, and it has nothing to do with weapon
familiarity.

**Comment by mugling** [comment]
I'm also a little reluctant to throw out the implementation I've written considering I called for opinions three times, most recently in October.

**Comment by Coolthulhu** [comment]
I only see conflict about the skill picked here and the uncraft being always 0 difficulty skill-less.
Changing those (and rebasing/merging master) would be enough to get it in, as far as I can tell.

Changing reloading to use gun skills could go to a separate issue/PR since there is no agreement about it, meaning it will hold up this PR otherwise.

**Comment by mugling** [comment]
Having written all of the prior code support and then called for comments on multiple occasions and finally stated an intention to proceed with an implementation there is the question of fairness.

The options here are:
* maintaining the status quo
* the implementation of (1) that I've written
* someone else writing an implementation (+/- new skills)

**Comment by kevingranade** [comment]
On Dec 3, 2016 11:34 AM, "mugling" <notifications@github.com> wrote:

I'm also a little reluctant to throw out the implementation I've written
considering I called for opinions three times, most recently in October.

I seriously feel bad about my unavailability, but at the same time it has
been unavoidable and it doesn't invalidate my points.

As Coolthulu points out, there is no controversy about the bulk of the
code, it's a simple matter of reverting the skill usage to fabrication if
you don't want to push forward to adding a new skill and reversing the 0
cost disassembly change.

**Comment by mugling** [comment]
> I seriously feel bad about my unavailability, but at the same time it has been unavoidable and it doesn't invalidate my points.

Belated congratulations - you shouldn't feel bad about that at all.

No it doesn't invalidate your points, but after a developer calls consensus on a number of occasions then it's reasonable to proceed with an implementation and expect it to be merged. There isn't otherwise any reason to seek consensus in the first place.

I think you'd agree I've made a lot of improvements to this area of the code base. It's not always easy to work on because it attracts a lot of lengthy opinion pieces. This often leads to endless discussion for no meaningful progress. The whole purpose of this issue was to avoid such deadlock.

Reloading recipes should depend on weapon skill so as to provide some reasonable progression otherwise we needn't have many of the interim recipes at all (why craft `9mmP` when you can craft `9mmP2`). I don't however want to reopen another cycle of discussion.

Open source developers typically only write the code they want to work on. Expecting them to work on anything else, especially ideas they are opposed to, isn't particularly realistic.

This brings us back to:
* maintaining the status quo
* the implementation of (1) that I've written
* someone else writing an implementation (+/- new skills)

**Comment by mugling** [comment]
Bump as `0.D` milestone

**Comment by kevingranade** [comment]
You're attempting to push through a redefinition of what these skills mean (a redefinition no core developer agrees with) as part of a mechanical change, that's not ok.
As it's clear you aren't going to address the problems, I'm closing the pr.

**Comment by mugling** [comment]
There were calls for consensus here which did appear to exist ~20 days ago prior to your return. The options are now reduced to you writing an implementation yourself or not releasing `0.D`.

**Comment by kevingranade** [comment]
> after a developer calls consensus on a number of occasions then it's reasonable to proceed with an implementation and expect it to be merged.

That's not how it works, it gets merged if a merger merges it.  All of the process that happens before a merge is coordination to try and find issues with the pr earlier in the development process.  If that process fails like it did here, it has no bearing on the decision to merge or not.

**Comment by WizardOfOoo** [comment]
Fourth Option: Change the skill back to 'gun' and keep virtually everything else, make the proposed skill changes a different issue.  If the goal is 0.D, then this change moves us forward.
If we close this one, somebody (me maybe) can work on [http://github.com/CleverRaven/Cataclysm-DDA/issues/16849]

**Comment by kevingranade** [comment]
I merged the acceptable parts of this with the skill changes myself and Coolthulu objected to earlier today, github lost track of it because I cherry-picked some of the changes instead of merging them.

**Pull request closed without merging**